libkbfs/delayed_cancellation.go

```diff
@@ -0,0 +1,224 @@
+// functions attached to ctx are run on the new new context. In addition, the
+// new context is made "cancellation delayable". That is, it disconnects the cancelFunc
+// from ctx, and watch for the cancellation. When cancellation happens, it
+// checks if delayed cancellation is enabled for the associated context. If so,
+// it waits until it's disabled before cancelling the new context. This
+// provides a hacky way to allow finer control over cancellation.
+//
+// Note that, it's important to call context.WithCancel (or its friends) before
+// this function if those cancellations need to be controllable ("cancellation
+// delayable"). Otherwise, the new cancelFunc is inheriently NOT ("cancellation
+// delayable").
+//
+// If this function is called, it is caller's responsibility to either 1)
+// cancel ctx (the context passed in); or 2) call CleanupCancellationDelayer;
+// when operations associated with the context is done. Otherwise it leaks go
+// routines!
+func NewContextWithCancellationDelayer(
+	ctx context.Context) (newCtx context.Context, err error) {
+	v := ctx.Value(CtxCancellationDelayerKey)
+	if v != nil {
+		if _, ok := v.(*cancellationDelayer); ok {
+			return nil, ContextAlreadyHasCancellationDelayerError{}
+		}
+		return nil, NoCancellationDelayerError{}
+	}
+
+	if newCtx, err = NewContextWithReplayFrom(ctx); err != nil {
+		return nil, err
+	}
+	c := newCancellationDelayer()
+	newCtx = NewContextReplayable(newCtx,
+		func(ctx context.Context) context.Context {
+			return context.WithValue(ctx, CtxCancellationDelayerKey, c)
+		})
+	newCtx, cancel := context.WithCancel(newCtx)
+	go func() {
+		select {
+		case <-ctx.Done():
+		case <-c.done:
+		}
+		c.L.Lock()
+		for c.delayEnabled {
+			c.Wait()
+		}
+		c.canceled = true
+		c.L.Unlock()
+		cancel()
+	}()
+	return newCtx, nil
+}
+
+// EnableDelayedCancellationWithGracePeriod can be called on a "cancellation
+// delayable" context produced by NewContextWithCancellationDelayer, to enable
+// delayed cancellation for ctx. This is useful to indicate that the
+// operation(s) associated with the context has entered a critical state, and
+// it should not be canceled until after timeout or CleanupCancellationDelayer
+// is called.
+func EnableDelayedCancellationWithGracePeriod(ctx context.Context, timeout time.Duration) error {
+	if c, ok := ctx.Value(CtxCancellationDelayerKey).(*cancellationDelayer); ok {
+		c.L.Lock()
+		defer c.L.Unlock()
+		if c.canceled || (c.timer != nil && !c.timer.Stop()) {
+			// Too late! The context is already canceled or the timer for
+			// disableDelay is already fired.
+			return context.Canceled
+		}
+		if !c.delayEnabled {
+			c.delayEnabled = true
+			c.Broadcast()
+		}
+		c.timer = time.AfterFunc(timeout, c.disableDelay)
+		return nil
+	}
+	return NoCancellationDelayerError{}
+}
+
+func (c *cancellationDelayer) disableDelay() {
+	c.L.Lock()
+	defer c.L.Unlock()
+	c.disableDelayLocked()
+}
+
+func (c *cancellationDelayer) disableDelayLocked() {
```

Comment:
I don't see another caller of this function.  Can you combine it with the one above?